File: backend/services/media_service.js
import { randomUUID } from "node:crypto";
import { query } from "../utils/errors.js";

// Image bytes go straight to Storage; Express only issues upload permission.
export async function createUpload(db, userId, extension) {
  const path = `${userId}/${randomUUID()}.${extension}`;
  const data = await query(
    db.storage.from("journal-media").createSignedUploadUrl(path),
  );
  return { path, signedUrl: data.signedUrl, token: data.token };
}
// Storage RLS permits owned images and images referenced by a published stop.
// Expiring links are for rendering only; save the permanent path in the journal.
export function readMedia(db, path) {
  return query(db.storage.from("journal-media").createSignedUrl(path, 300));
}

File: backend/services/media_service.py
"""Create permissions for direct uploads instead of routing image bytes through Python."""

from uuid import uuid4
from utils.errors import storage_call


def create_upload(db, user_id, extension):
    path = f"{user_id}/{uuid4()}.{extension}"
    result = storage_call(
        lambda: db.storage.from_("journal-media").create_signed_upload_url(path)
    )
    return {"path": path, "signedUrl": result["signed_url"], "token": result["token"]}


def read_media(db, path):
    # Storage RLS allows own photos and photos explicitly shared in a published post.
    result = storage_call(
        lambda: db.storage.from_("journal-media").create_signed_url(path, 300)
    )
    return {"signedUrl": result["signedURL"]}
